analytics/src/analytics_android.cc

```diff
@@ -355,6 +355,108 @@ void AddToBundle(JNIEnv* env, jobject bundle, const char* key, int64_t value) {
+  env->CallVoidMethod(bundle,
+                      util::bundle::GetMethodId(util::bundle::kPutBundle),
+                      key_string, inner_bundle);
+  util::CheckAndClearJniExceptions(env);
+  env->DeleteLocalRef(key_string);
+}
+
+// Declared here so that it can be used, defined below.
+jobject MapToBundle(JNIEnv* env, const std::map<Variant, Variant>& map);
+
+// Converts the given vector into a Java ArrayList. It is up to the
+// caller to delete the local reference when done.
+jobject VectorToArrayList(JNIEnv* env, const std::vector<Variant>& vector) {
+  jobject arraylist = env->NewObject(
+      util::array_list::GetClass(),
+      util::array_list::GetMethodId(util::array_list::kConstructor));
+
+  for (const Variant& element : vector) {
+    if (element.is_map()) {
+      jobject bundle = MapToBundle(env, element.map());
+      env->CallBooleanMethod(
+          arraylist, util::array_list::GetMethodId(util::array_list::kAdd),
+          bundle);
+      util::CheckAndClearJniExceptions(env);
+      env->DeleteLocalRef(bundle);
+    } else {
+      LogError("VectorToArrayList: Unsupported type (%s) within vector.",
+               Variant::TypeName(element.type()));
+    }
+  }
+  return arraylist;
+}
+
+// Converts and adds the Variant to the given Bundle.
+bool AddVariantToBundle(JNIEnv* env, jobject bundle, const char* key,
+                        const Variant& value) {
+  if (value.is_int64()) {
+    AddToBundle(env, bundle, key, value.int64_value());
+  } else if (value.is_double()) {
+    AddToBundle(env, bundle, key, value.double_value());
+  } else if (value.is_string()) {
+    AddToBundle(env, bundle, key, value.string_value());
+  } else if (value.is_bool()) {
+    // Just use integer 0 or 1.
+    AddToBundle(env, bundle, key,
+                value.bool_value() ? static_cast<int64_t>(1L)
+                                   : static_cast<int64_t>(0L));
+  } else if (value.is_null()) {
+    // Just use integer 0 for null.
+    AddToBundle(env, bundle, key, static_cast<int64_t>(0L));
+  } else if (value.is_vector()) {
+    jobject arraylist = VectorToArrayList(env, value.vector());
+    AddArrayListToBundle(env, bundle, key, arraylist);
+    env->DeleteLocalRef(arraylist);
+  } else if (value.is_map()) {
+    jobject inner_bundle = MapToBundle(env, value.map());
+    AddBundleToBundle(env, bundle, key, inner_bundle);
+    env->DeleteLocalRef(inner_bundle);
+  } else {
+    // A Variant type that couldn't be handled was passed in.
+    return false;
+  }
+  return true;
+}
+
+// Converts the given map into a Java Bundle. It is up to the caller
+// to delete the local reference when done.
+jobject MapToBundle(JNIEnv* env, const std::map<Variant, Variant>& map) {
+  jobject bundle =
+      env->NewObject(util::bundle::GetClass(),
+                     util::bundle::GetMethodId(util::bundle::kConstructor));
+  for (const auto& pair : map) {
+    // Only add elements that use a string key
```

Comment:
Just making sure there are no numeric keys allowed?

Reply:
Yep, it is currently only strings, and I doubt that'll change any time soon.